src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -2293,15 +2293,24 @@ Status Session::receive_ident_message(SaltedFileIdent client_file_ident)
     // if a client reset happens, it will take care of setting the file ident
     // and if not, we do it here
     bool did_client_reset = false;
-    try {
-        did_client_reset = client_reset_if_needed();
-    }
-    catch (const std::exception& e) {
+
+    auto on_session_error = [this](const auto& e, void* error_code = nullptr) {
         auto err_msg = util::format("A fatal error occurred during client reset: '%1'", e.what());
+        SessionErrorInfo err_info(Status{ErrorCodes::AutoClientResetFailed, err_msg}, IsFatal{true}, error_code);
         logger.error(err_msg.c_str());
-        SessionErrorInfo err_info(Status{ErrorCodes::AutoClientResetFailed, err_msg}, IsFatal{true});
         suspend(err_info);
         return Status::OK();
+    };
+
+    try {
+        did_client_reset = client_reset_if_needed();
+    }
+    catch (const UserCodeCallbackError& e) {
```

Comment:
I don't think stuff in client_impl_base should have to know about the details of the CAPI's error handling conventions. can't we push this all down into the C API?

Reply:
If the exception happens in usercode, the SDK needs to set some extra data there, to feed into Session::suspend() ... however maybe all we need to do is to construct a C++ exception to rethrow with the message of the original exception that was triggered in usercode. This could simplify the code...

Reply:
Actually this catch block is not needed.. scratch my previous comment @jbreams ...